File: backend/dataset/maxPathSum/maxPathSum_2.py
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def check(node: TreeNode):
            (lps, ls) = check(node.left) if node.left else (0, float('-inf'))
            (rps, rs) = check(node.right) if node.right else (0, float('-inf'))
            return max(lps + node.val, rps + node.val, node.val), max(ls, rs, lps + node.val, rps + node.val, lps + rps + node.val, node.val)
        return check(root)[1]

File: backend/CDataset/maxPathSum/maxPathSum_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    class TreeNode{
        public:
        TreeNode* left;
        TreeNode* right;
        int val;
    };
    int ans=INT_MIN;
    int solve(TreeNode* root)
    {
        if(root==NULL) return 0;
        int ls = max(0,solve(root->left));
        int rs = max(0,solve(root->right));
        ans = max(ans,root->val+ls+rs);
        return root->val+max(ls,rs);
    }
    int maxPathSum(TreeNode* root) {
        solve(root);
        return ans;
    }
};
